Fix debug assertion in EngineBuffer

uklotzde opened the pull request:
The debug assertion failed in EngineBufferTest.SlowRubberBand:
https://bugs.launchpad.net/mixxx/+bug/1580895

The debug assertion is invalid and just catches rounding errors when converting between samples and frames. All (sub-)position calculations in Mixxx should be done in frames instead of samples. This PR only fixes the EngineBufferScale calculations.

Unfortunately I noticed too late that @daschuer was already working on this. I expect serious merge conflicts with this branch: https://github.com/daschuer/mixxx/commits/positionrounding2

Do whatever you want with this PR. I just want to get this annoying debug assertion out of my way that no one seemed to care about for months...

uklotzde committed:
Disable and document violated debug assertion in EngineBuffer

The debug assertion failed in EngineBufferTest.SlowRubberBand:
https://bugs.launchpad.net/mixxx/+bug/1580895

Without the debug assertion the test completes successfully.

uklotzde committed:
Fix calculation of samplesRead for edge cases in EngineBufferScale

Wrong calculation has been introduced by the following commit:
a56c50dd3d967c1be4ad3683c682c1d925449792

uklotzde committed:
Enable debug_assertions_fatal for Travis builds

Needed for early detection of violated debug assertions.

uklotzde committed:
Re-enable debug assertion

uklotzde committed:
Delete unnecessary cast from double to int

uklotzde committed:
Perform scaler calculations in frames instead of samples

uklotzde committed:
Eliminate implicit assumptions about audio signals in scaler code

Well, as far as possible without rewriting large parts of the code.

daschuer commented:
Hi @uklotzde, thank you for your PR. It fits pretty nice to my branch. I have just re-based it onto yours and there are only some minor conflicts. You have actually worked on my todos :-)

daschuer commented on src/engine/enginebuffer.cpp:
```diff
@@ -41,8 +41,18 @@
+
+const SINT kSamplesPerFrame = 2; // Engine buffer uses Stereo frames only
+
+inline void assertValidPlayPosition(double playPosition) {
```
assertPlayPositionAtFrameStart

We may also consider to make kSamplesPerFrame a parameter and move it to util/sample.h

daschuer commented on src/engine/enginebufferscale.h:
```diff
@@ -65,17 +48,27 @@ class EngineBufferScale : public QObject {
+    // The actual number of frames copied to the output buffer is always
+    // an integer value, while the virtual number of read frames might be
+    // partial number!
+    virtual double getScaledSampleFrames(
```
sample frames is a new name. 
Since the parameter is a pointer to samples and the iBufferSize is a sample count,
getScaledSamples() is correct if we look to the output parameter. 
If we look on the return value it returns the consumed frames from unscaled input buffer. 
getReadFrames(); 
On the other hand this is no classic getter ... mmm

How about: scaleFrames()

daschuer commented on src/engine/enginebufferscale.h:
```diff
@@ -65,17 +48,27 @@ class EngineBufferScale : public QObject {
     virtual void clear() = 0;
     // Scale buffer
-    virtual double getScaled(CSAMPLE* pOutput, const int iBufferSize) = 0;
+    // Returns the virtual number of sample frames that have been read.
```
The term virtual is misleading here since the frames of the file are real.  

"Returns the number of frames that have bean read from the unscaled input buffer. ..."

daschuer commented on src/engine/enginebuffer.cpp:
```diff
@@ -448,6 +458,7 @@ void EngineBuffer::readToCrossfadeBuffer(const int iBufferSize) {
 void EngineBuffer::setNewPlaypos(double newpos) {
     //qDebug() << m_group << "engine new pos " << newpos;
 
+    assertValidPlayPosition(newpos);
```
It is fully valid to seek to a fractional frame.

uklotzde replied:
Are you sure? There is another check including a correction of the parameter 
value in EngineBuffer::doSeekPlayPos()!

What's the difference between setNewPlaypos() and doSeekPlayPos()?

On 27.05.2016 23:47, Daniel Schürmann wrote:
>
> In src/engine/enginebuffer.cpp 
> <https://github.com/mixxxdj/mixxx/pull/954#discussion_r64970724>:
>
> > @@ -448,6 +458,7 @@ void EngineBuffer::readToCrossfadeBuffer(const int iBufferSize) {
> >  void EngineBuffer::setNewPlaypos(double newpos) {
> >      //qDebug() << m_group << "engine new pos " << newpos;
> >
> > +    assertValidPlayPosition(newpos);
>
> It is fully valid to seek to a fractional frame.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/954/files/78c1ea84111198fd32ecdd84a6ed448fd103f145#r64970724>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/ABZ-atoYsxbJH8uMjtCzXWSocXZ_tsdyks5qF2aBgaJpZM4ImsNg>.
>

uklotzde replied:
My last commit deletes all rounding of play positions to frame boundaries

On 28.05.2016 12:20, Uwe Klotz wrote:
> Are you sure? There is another check including a correction of the 
> parameter value in EngineBuffer::doSeekPlayPos()!
>
> What's the difference between setNewPlaypos() and doSeekPlayPos()?
>
> On 27.05.2016 23:47, Daniel Schürmann wrote:
>>
>> In src/engine/enginebuffer.cpp 
>> <https://github.com/mixxxdj/mixxx/pull/954#discussion_r64970724>:
>>
>> > @@ -448,6 +458,7 @@ void EngineBuffer::readToCrossfadeBuffer(const int iBufferSize) {
>> >  void EngineBuffer::setNewPlaypos(double newpos) {
>> >      //qDebug() << m_group << "engine new pos " << newpos;
>> >
>> > +    assertValidPlayPosition(newpos);
>>
>> It is fully valid to seek to a fractional frame.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/mixxxdj/mixxx/pull/954/files/78c1ea84111198fd32ecdd84a6ed448fd103f145#r64970724>, 
>> or mute the thread 
>> <https://github.com/notifications/unsubscribe/ABZ-atoYsxbJH8uMjtCzXWSocXZ_tsdyks5qF2aBgaJpZM4ImsNg>.
>>
>

daschuer commented on src/engine/enginebufferscalelinear.cpp:
```diff
@@ -88,45 +79,47 @@ double EngineBufferScaleLinear::getScaled(CSAMPLE* pOutput,
         m_dRate = rate_add_new;
         // pass the address of the sample at the halfway point
-        samples_read += do_scale(&pOutput[buf_size / 2], buf_size / 2);
+        frames_read += do_scale(&pOutputBuffer[getAudioSignal().samples2frames(iOutputBufferSize)], getAudioSignal().samples2frames(iOutputBufferSize));
```
loooooooooooooooooooooooooooooooooooooooooooooooooooong line ;-)

daschuer commented on src/engine/enginebufferscalest.cpp:
```diff
@@ -120,51 +98,49 @@ void EngineBufferScaleST::setScaleParameters(double base_rate,
+        SINT iOutputBufferSize) {
+    if (m_dBaseRate == 0.0 || m_dTempoRatio == 0.0 || m_dPitchRatio == 0.0) {
+        SampleUtil::clear(pOutputBuffer, iOutputBufferSize);
+        return getAudioSignal().samples2frames(iOutputBufferSize);
```
return 0.0

daschuer commented on src/engine/enginebufferscalerubberband.cpp:
```diff
@@ -154,34 +147,31 @@ void EngineBufferScaleRubberBand::deinterleaveAndProcess(
+        SINT iOutputBufferSize) {
+    if (m_dBaseRate == 0.0 || m_dTempoRatio == 0.0) {
+        SampleUtil::clear(pOutputBuffer, iOutputBufferSize);
+        return getAudioSignal().samples2frames(iOutputBufferSize);
```
return 0.0

rryan commented:
Big +1 to changing things to work with frames instead of samples -- fractional frames are indeed fine in theory but working in samples made the logic too tricky for avoiding overflow.

I think we can all agree this should not go in 2.1 though?

daschuer commented on src/engine/enginebufferscalerubberband.h:
```diff
@@ -13,37 +14,40 @@ class ReadAheadManager;
+    void initRubberBand();
+
     void deinterleaveAndProcess(const CSAMPLE* pBuffer, size_t frames, bool flush);
     size_t retrieveAndDeinterleave(CSAMPLE* pBuffer, size_t frames);
```
SINT

daschuer commented on src/engine/enginebufferscalest.cpp:
```diff
@@ -120,51 +98,49 @@ void EngineBufferScaleST::setScaleParameters(double base_rate,
     bool last_read_failed = false;
     while (remaining_frames > 0) {
-        unsigned int received_frames = m_pSoundTouch->receiveSamples(
+        SINT received_frames = m_pSoundTouch->receiveSamples(
```
receiveSamples() returns an UINT

daschuer commented:
> I think we can all agree this should not go in 2.1 though?

Yes, This can go to master once we have a 2.1 branch, which is needed soon to merge upcoming the GSoC stuff.

uklotzde committed:
Rename function for verifying valid play positions

uklotzde committed:
Consistently check for valid seek/play positions

uklotzde committed:
Delete another implicit assumption about number of channels

uklotzde committed:
Fix split buffer scaling in linear scaler

uklotzde committed:
Use SINT for all sample/frame index and length calculations

uklotzde committed:
Fix and explain the scaling result for edge cases

uklotzde committed:
Rename and document getScaledSampleFrames()

uklotzde committed:
Do not round play positions to frame boundaries

The debug assertions failed during tests, so we should keep everything
in floating point here.

daschuer commented:
LGTM, Thank you. 

I have rebased my https://github.com/daschuer/mixxx/tree/positionrounding2 on this and added a test for fractional frame loops. How should we continue?
It would be best to merge this to a non 2.1 master then I can issue my PR as usual.  
Or should I issue a PR against master including this commits, or a PR against this branch?

uklotzde committed:
Merge branch 'master' into EngineBufferDebugAssertion

daschuer commented:
Anny comments to post above? 

@rryan: How about to cut out a 2.1 branch now?

sblaisot commented:
@daschuer there a some PR to merge before cutting a 2.1 branch I think. In particular I would like to have PR #862 merged and tested before cutting a 2.1 branch, otherwise windows installer switch to wix will be delayed to 2.2 or 3.0.

IMHO a 2.1 branch needs a working build server to release rolling betas during the 2.1 QA phase.

daschuer commented:
Unfortunately this one is starting to rod. We should avoid slow down our selves because not having a 2.1 branch.  Next week we might whant to review GSoC midterm results, not ready for 2.1.  

Unfortunately I cannot give a final LGTM for https://github.com/mixxxdj/mixxx/pull/862 because I have no windows build environment.  Can't we just merge it and improve it later if required. 
It should be not a big deal to open a  #862 again for a new 2.1. on the other hand.  

Any ideas?

daschuer commented:
Since we did not manage to cut a branch since a month, how about merging this now?

daschuer commented:
Since we can declare any commit as 2.1 branch, I will merge it now. 
Thank you @uklotzde for all the work you did here and in other PRs.
